rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,321 @@
+# StableHLO Bounded Dynamism
+
+[StableHLO](https://github.com/openxla/stablehlo) is an operation set that
+expresses ML computations. It has been originally bootstrapped from the [MHLO
+dialect](https://github.com/tensorflow/mlir-hlo#meta-hlo-dialect-mhlo),
+including inheriting the type and some ops. This RFC aims to describe the
+current status and rationale for bounded dynamism constructs in StableHLO and
+provides recommendations to StableHLO producers and consumers. In particular,
+this RFC doesn’t propose any further changes to the current state but this RFC
+should still be used to revisit those decisions given that those weren't
+discussed publicly.
+
+Bounded dynamism allows programs to represent the maximum runtime size that a
+particular dynamic dimension of a tensor can have. This makes it possible to run
+such programs on hardware that don't support dynamic tensors but could support
+it if the upper bounds of tensor dimensions are known at the compilation time.
+With bounded dynamism, real time inference systems don't need to wait for
+accumulation up to a particular batch size on such hardware. Bounded dynamism
+also makes it possible to support programs whose intermediate tensor shapes
+depend on the inputs. For example,
+[`stablehlo.dynamic_broadcast_in_dim`](https://github.com/openxla/stablehlo/blob/ff55f9346d54e9e38de807a79f8ae03faffda274/stablehlo/dialect/StablehloOps.td#L1838)
+op but with statically known upper bounds of `output_dimensions` operand. Even
+on hardware that supports dynamic tensors, bounded dynamism can open up
+performance optimizations opportunities.
+
+# Recommendations
+
+## StableHLO Producers
+
+* Producers should use the bounded tensor type representation as described in
+  P1. Function arguments of public functions should be bounded.
+* Producers are encouraged to use the unbounded dynamic operations as described
+  in P4. Result types of these ops aren't required to have a bounded type.
```

Comment:
Renamed this section to '(P4) Aspirational: Migration to unbounded dynamism' as discussed offline.

Reply:
Since this section is aspirational (i.e. talks about something that may be implemented in the future), it is not necessarily a great fit for recommendations (i.e. something that producers should be doing right now).

Reply:
Removed as per offline discussion.